src/HexagonOffload.cpp

```diff
@@ -0,0 +1,322 @@
+                arg_sizes.push_back(Expr((uint64_t) sizeof(buffer_t*)));
+                arg_ptrs.push_back(Variable::make(type_of<buffer_t *>(), i.first + ".buffer"));
+                int flags = 0;
+                if (i.second.read) flags |= 0x1;
+                if (i.second.write) flags |= 0x2;
+                arg_flags.push_back(flags);
+            }
+            for (const auto& i : c.vars) {
+                Expr arg = Variable::make(i.second, i.first);
+                Expr arg_ptr = Call::make(type_of<void *>(), Call::make_struct, {arg}, Call::Intrinsic);
+
+                // sizeof(scalar-type) will always fit into int32
+                arg_sizes.push_back(Expr((uint64_t) i.second.bytes()));
+                arg_ptrs.push_back(arg_ptr);
+                arg_flags.push_back(0x0);
+            }
+
+            // The argument list is terminated with an argument of size 0.
+            arg_sizes.push_back(Expr((uint64_t) 0));
+
+            std::string pipeline_name = hex_name + "_argv";
+            std::vector<Expr> params;
+            params.push_back(module_state());
+            params.push_back(pipeline_name);
+            params.push_back(state_var_ptr(hex_name, type_of<int>()));
+            params.push_back(Call::make(type_of<size_t*>(), Call::make_struct, arg_sizes, Call::Intrinsic));
+            params.push_back(Call::make(type_of<void**>(), Call::make_struct, arg_ptrs, Call::Intrinsic));
+            params.push_back(Call::make(type_of<int*>(), Call::make_struct, arg_flags, Call::Intrinsic));
+
+            stmt = call_extern_and_assert("halide_hexagon_run", params);
+
+        } else {
+            IRMutator::visit(loop);
+        }
+    }
+
+    void visit(const Let *op) {
+        if (op->value.type() == Int(32)) {
+            alignment_info.push(op->name, modulus_remainder(op->value, alignment_info));
+        }
+
+        IRMutator::visit(op);
+
+        if (op->value.type() == Int(32)) {
+            alignment_info.pop(op->name);
+        }
+    }
+
+    void visit(const LetStmt *op) {
+        if (op->value.type() == Int(32)) {
+            alignment_info.push(op->name, modulus_remainder(op->value, alignment_info));
+        }
+
+        IRMutator::visit(op);
+
+        if (op->value.type() == Int(32)) {
+            alignment_info.pop(op->name);
+        }
+    }
+
+public:
+    InjectHexagonRpc(const Target &target) : device_code("hexagon", target) {}
+
+    Stmt inject(Stmt s) {
+        s = mutate(s);
+
+        // Skip if there are no device kernels.
+        if (device_code.functions().empty()) {
+            return s;
+        }
+
+        // Compile the device code.
+        TemporaryFile tmp_bitcode("hex", ".bc");
+        TemporaryFile tmp_shared_object("hex", ".so");
+        debug(1) << "Hexagon device code module: " << device_code << "\n";
+        device_code.compile(Outputs().bitcode(tmp_bitcode.pathname()));
+
+        string hex_command = "${HL_HEXAGON_TOOLS}/bin/hexagon-clang ";
```

Comment:
I'm pretty sure that this whole section (i.e. shelling out to hexagon-clang) requires a TODO comment...